rosbag2_storage/include/rosbag2_storage/metadata_io.hpp

```diff
@@ -12,25 +12,30 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#ifndef ROSBAG2_STORAGE__STORAGE_INTERFACES__BASE_INFO_INTERFACE_HPP_
-#define ROSBAG2_STORAGE__STORAGE_INTERFACES__BASE_INFO_INTERFACE_HPP_
+#ifndef ROSBAG2_STORAGE__METADATA_IO_HPP_
+#define ROSBAG2_STORAGE__METADATA_IO_HPP_
 
-#include "rosbag2_storage/bag_info.hpp"
+#include <string>
+
+#include "rosbag2_storage/bag_metadata.hpp"
+#include "rosbag2_storage/topic_with_type.hpp"
 #include "rosbag2_storage/visibility_control.hpp"
 
 namespace rosbag2_storage
 {
-namespace storage_interfaces
-{
 
-class ROSBAG2_STORAGE_PUBLIC BaseInfoInterface
+class MetadataIo
 {
 public:
-  virtual ~BaseInfoInterface() = default;
-  virtual BagInfo info() = 0;
+  static constexpr const char * const metadata_filename = "metadata.yaml";
```

Comment:
is the user ever able to change this name? What's happening if a user changes this file to another name?

Reply:
At the moment, changing this name is not supported. As we discussed, this would require the user to also be able to give a path not just to the directory but to the metadata file itself. The existence of the metadata.yaml file is an implementation detail and i won't exist for old bagfiles, for instance.

Exposing this to the user by allowing to specify the path to the yaml file directly might (as discussed) be a feature, but I'd suggest we open a ticket for now.

Reply:
At the moment, changing this name is not supported. As we discussed, this would require the user to also be able to give a path not just to the directory but to the metadata file itself. The existence of the metadata.yaml file is an implementation detail and i won't exist for old bagfiles, for instance.

Exposing this to the user by allowing to specify the path to the yaml file directly might (as discussed) be a feature, but I'd suggest we open a ticket for now.

Reply:
Not at the moment. As discussed, with the current changes the user enters a path to a directory containing this file. For simple use cases, this should be sufficient: Treat the directory as bagfile and hide implementation details (e.g. the metadata file, database, etc.) from the user. 

For more complicated use cases, if you want to modify the folder structure, you can already do that by also changing the content metadata file, but until it is possible to specify the path to a yaml file containing the information, it is impossible to rename the metadata file.

We suggest to open a new issue to allow the user to enter a path to the yaml file directly as this is a separate feature. At that point, it would be simple to support renaming the metadata file.

Reply:
At the moment, it is only possible to hand the name of the directory to `ros2 bag record` (as discussed). For simple use-cases, I think this is sufficient: We think of the folder as bagfile and hide the implementation details (database, metadata yaml file) from the user.

For more complicated use-cases it is already possible to restructure the folder completely by adapting the yaml file. As long as it's impossible to hand the path to the yaml file directly to ros2 bag, it's impossible to rename the yaml file.

We suggest to open an issue to allow specifying the path to a yaml file instead of the path to a folder containing a yaml file. With this change, it is possible to support renaming the metadata yaml file.

Reply:
https://github.com/ros2/rosbag2/issues/49